File: priv/assets/js/rolling_number.js
// RollingNumber — animate a number to a new value (Mantine RollingNumber parity).
//
// On mount it rolls from 0 up to `data-value`; whenever LiveView re-renders with a new `data-value`
// it animates from the current value to the new one (requestAnimationFrame, ease-out). `data-duration`
// (ms, default 800) sets the speed and `data-locale` the thousands grouping. Respects
// `prefers-reduced-motion` by jumping straight to the value.

const RollingNumber = {
  mounted() {
    this.current = 0;
    this.animateTo(this.target());
  },

  updated() {
    this.animateTo(this.target());
  },

  target() {
    return Number(this.el.getAttribute("data-value")) || 0;
  },

  format(n) {
    const locale = this.el.getAttribute("data-locale") || undefined;
    return Math.round(n).toLocaleString(locale);
  },

  animateTo(target) {
    const from = this.current;
    if (from === target) {
      this.el.textContent = this.format(target);
      return;
    }

    const reduce =
      window.matchMedia && window.matchMedia("(prefers-reduced-motion: reduce)").matches;
    const dur = reduce ? 0 : Number(this.el.getAttribute("data-duration")) || 800;

    if (dur <= 0) {
      this.current = target;
      this.el.textContent = this.format(target);
      return;
    }

    const start = performance.now();
    cancelAnimationFrame(this._raf);

    const tick = (now) => {
      const t = Math.min(1, (now - start) / dur);
      const eased = 1 - Math.pow(1 - t, 3);
      const val = from + (target - from) * eased;
      this.el.textContent = this.format(val);
      if (t < 1) {
        this._raf = requestAnimationFrame(tick);
      } else {
        this.current = target;
        this.el.textContent = this.format(target);
      }
    };

    this._raf = requestAnimationFrame(tick);
  },

  destroyed() {
    cancelAnimationFrame(this._raf);
  },
};

export default RollingNumber;

File: development/lib/development_web/components/headless/rolling_number.ex
defmodule DevelopmentWeb.Components.Headless.RollingNumber do
  @moduledoc """
  Headless **rolling number** — animate a number to its value (Mantine RollingNumber parity).

  The `RollingNumber` engine rolls from 0 up to `value` on mount and animates to the new number
  whenever LiveView re-renders with a different `value` (requestAnimationFrame, ease-out). `duration`
  (ms) sets the speed, `locale` the thousands grouping; `prefers-reduced-motion` jumps straight to the
  value. The static `value` is rendered as fallback text for no-JS / screen readers.

  Ships **no** styling — style via `chelekom-rolling-number`.

  **Documentation:** https://mishka.tools/chelekom/docs/headless/rolling_number
  """
  use Phoenix.Component

  @doc type: :component
  attr :id, :string, required: true, doc: "Unique id (carries the RollingNumber hook)"
  attr :value, :integer, required: true, doc: "The target number"
  attr :duration, :integer, default: 800, doc: "Animation duration in ms"
  attr :locale, :string, default: nil, doc: "BCP-47 locale for thousands grouping"
  attr :class, :any, default: nil, doc: "Extra classes"
  attr :rest, :global

  def rolling_number(assigns) do
    ~H"""
    <span
      id={@id}
      phx-hook="RollingNumber"
      data-part="root"
      data-value={@value}
      data-duration={@duration}
      data-locale={@locale}
      class={["chelekom-rolling-number", @class]}
      {@rest}
    >{@value}</span>
    """
  end
end
